smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitorTest.kt

```diff
@@ -5,91 +5,261 @@
 
 package software.amazon.smithy.kotlin.codegen.rendering.waiters
 
+import io.kotest.matchers.string.shouldEndWith
+import software.amazon.smithy.codegen.core.CodegenException
 import software.amazon.smithy.jmespath.JmespathExpression
-import software.amazon.smithy.kotlin.codegen.KotlinCodegenPlugin
 import software.amazon.smithy.kotlin.codegen.core.KotlinWriter
-import software.amazon.smithy.kotlin.codegen.loadModelFromResource
-import software.amazon.smithy.kotlin.codegen.model.expectShape
 import software.amazon.smithy.kotlin.codegen.test.TestModelDefault
-import software.amazon.smithy.kotlin.codegen.test.createSymbolProvider
 import software.amazon.smithy.kotlin.codegen.test.stripCodegenPrefix
-import software.amazon.smithy.model.shapes.ShapeId
-import software.amazon.smithy.model.shapes.StructureShape
 import kotlin.test.Test
 import kotlin.test.assertEquals
+import kotlin.test.fail
 
 class KotlinJmespathExpressionVisitorTest {
-    val model = loadModelFromResource("waiter-tests.smithy")
-    val provider = KotlinCodegenPlugin.createSymbolProvider(model)
-    val inputShape = model.expectShape<StructureShape>(ShapeId.from("${TestModelDefault.NAMESPACE}#DescribeFooInput"))
-    val inputSymbol = provider.toSymbol(inputShape)
-    val outputShape = model.expectShape<StructureShape>(ShapeId.from("${TestModelDefault.NAMESPACE}#DescribeFooOutput"))
-    val outputSymbol = provider.toSymbol(outputShape)
+    @Test
+    fun testAndExpression() {
+        assertUnimplemented("foo && bar")
+    }
+
+    @Test
+    fun testComparatorExpression_Equality() {
+        assertVisit(
+            "foo == bar",
+            "comparison",
+            "val foo = it?.foo",
+            "val bar = it?.bar",
+            "val comparison = foo == bar",
+        )
+    }
+
+    @Test
+    fun testComparatorExpression_EqualityWithString() {
+        assertVisit(
+            "`\"foo\"` == bar",
+            "comparison",
+            "val string = \"foo\"",
+            "val bar = it?.bar",
+            "val comparison = string == bar?.toString()",
+        )
+
+        assertVisit(
+            "foo == `\"bar\"`",
+            "comparison",
+            "val foo = it?.foo",
+            "val string = \"bar\"",
+            "val comparison = foo?.toString() == string",
+        )
+    }
+
+    @Test
+    fun testComparatorExpression_EqualityWithTwoStrings() {
+        assertVisit(
+            "`\"foo\"` != `\"bar\"`",
+            "comparison",
+            "val string = \"foo\"",
+            "val string2 = \"bar\"",
+            "val comparison = string != string2",
+        )
+    }
+
+    @Test
+    fun testComparatorExpression_Inequality() {
+        assertVisit(
+            "foo <= bar",
+            "comparison",
+            "val foo = it?.foo",
+            "val bar = it?.bar",
+            "val comparison = if (foo == null || bar == null) null else foo <= bar",
```

Comment:
The return value here is `Boolean?`. There's no final checks anywhere else. Most everything should be safe because we treat all vals as nullable everywhere _except_ in the return of the `filter` call. That's missing a null-safe conclusion so I'll add one.

Reply:
I see now, it was related to my other more general question on the usage of the result of evaluating the expression.